compiler/src/dotty/tools/dotc/transform/PostTyper.scala

```diff
@@ -85,6 +92,69 @@ class PostTyper extends MacroTransform with IdentityDenotTransformer { thisPhase
     // TODO fill in
   }
 
+  private def normalizeTypeArgs(tree: TypeApply)(using Context): TypeApply = tree.tpe match
+    case pt: PolyType => // wait for more arguments coming
+      tree
+    case _ =>
+      def decompose(tree: TypeApply): (Tree, List[Tree]) = tree.fun match
+        case fun: TypeApply =>
+          val (tycon, args) = decompose(fun)
+          (tycon, args ++ tree.args)
+        case _ =>
+          (tree.fun, tree.args)
+      end decompose
+      def reorderArgs(pnames: List[Name], namedArgs: List[NamedArg], otherArgs: List[Tree]): List[Tree] = pnames match
+        case pname :: pnames1 =>
+          namedArgs.partition(_.name == pname) match
+            case (NamedArg(_, arg) :: _, namedArgs1) =>
+              arg :: reorderArgs(pnames1, namedArgs1, otherArgs)
+            case _ =>
+              val otherArg :: otherArgs1 = otherArgs
+              otherArg :: reorderArgs(pnames1, namedArgs, otherArgs1)
+        case nil =>
+          assert(namedArgs.isEmpty && otherArgs.isEmpty)
+          Nil
+      end reorderArgs
+
+      val (tycon, args) = decompose(tree)
+      tycon.tpe.widen match
+        case tp: PolyType if args.exists(isNamedArg) =>
+          val (namedArgs, otherArgs) = args.partition(isNamedArg)
+          val args1 = reorderArgs(tp.paramNames, namedArgs.asInstanceOf[List[NamedArg]], otherArgs)
+          TypeApply(tycon, args1).withSpan(tree.span).withType(tree.tpe)
+        case _ =>
+          tree
+  end normalizeTypeArgs
+
+  /** Scan a tree and check all the AppliedTypeTrees in it. */
+  def checkAppliedTypes(tree: Tree)(using Context): Unit = tree match
```

Comment:
I think this could be done using a TreeTraverser

Reply:
That simplifies things, thanks!